### PR TITLE
Add CIDER log middleware

### PR DESCRIPTION
**Pull request by r0man**
This PR adds the CIDER log middleware, now with logjam extracted into its own repository.

The initial PR was this one:
https://github.com/clojure-emacs/cider-nrepl/pull/773

And this for the CIDER side:
https://github.com/clojure-emacs/cider/pull/3352

Before submitting a PR make sure the following things have been done:

- [x] The commits are consistent with our [contribution guidelines](../blob/master/.github/CONTRIBUTING.md)
- [x] You've added tests to cover your change(s)
- [x] All tests are passing
- [x] The new code is not generating reflection warnings
- [x] You've updated the README (if adding/changing middleware)

Thanks!

### CODE REVIEW

**Comment by r0man** [comment]
@vemv @bbatsov Do we want to inline the logjam dependency or not? 

I gave it a try, but it looks like running the tests with the rewritten logjam library fail, because they reference the logjam library and not the re-written one. I also tried adding the non-rewritten logjam library to the test profile, but that also fails. 

If we want to inline I can take a look at somehow getting mr anderson to work with it, or get rid of the references in the tests (I'm using `framework/log` there to log test messages without having to deal with the specific implementation).

**Comment by bbatsov** [comment]
I'm fine with not inlining the dep - I don't expect any runtime conflicts with it. 

Btw, how are the tests passing on Clojure 1.8? I thought logjam required 1.9.

File: test/clj/cider/nrepl/middleware/log_test.clj [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+  `(doseq [framework# ~frameworks :let [~framework-sym framework#]]
+     (testing (format "Log framework %s" (:name framework#))
+       ~@body)))
+
+(defmacro with-appender
+  "Add an appender for `framework`, evaluate `body` and remove the appender."
+  [[framework appender options] & body]
+  `(let [framework# ~framework, appender# ~appender]
+     (add-appender framework# appender# ~options)
+     (try ~@body (finally (remove-appender framework# appender#)))))
```
[the remark this answers is not in the document]
Comment (r0man) [reply test/clj/cider/nrepl/middleware/log_test.clj:89]: Fixed

File: test/clj/cider/nrepl/middleware/log_test.clj [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+(deftest test-threads
+  (with-each-framework [framework (frameworks)]
+    (with-appender [framework appender]
+      (framework/log framework {:message "a-1"})
+      (let [response (session/message
+                      {:op "cider/log-threads"
+                       :framework (:id framework)
+                       :appender (:id appender)})]
+        (is (= #{"done"} (:status response)))
+        (let [threads (:cider/log-threads response)]
```
[the remark this answers is not in the document]
Comment (r0man) [reply test/clj/cider/nrepl/middleware/log_test.clj:346]: Fixed

File: test/clj/cider/nrepl/middleware/log_test.clj [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+          (is (every? keyword? (keys threads)))
```
[the remark this answers is not in the document]
Comment (r0man) [reply test/clj/cider/nrepl/middleware/log_test.clj:347]: Fixed

File: test/clj/cider/nrepl/middleware/log_test.clj [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+    (testing "remove unregistered appender"
+      (let [response (session/message
+                      {:op "cider/log-remove-appender"
+                       :framework (:id framework)
+                       :appender "unknown"})]
+        (is (= #{"cider/log-remove-appender-error" "done"} (:status response)))))
+    (testing "remove registered appender"
+      (let [appender (add-appender framework appender)
+            appender' (remove-appender framework appender)]
+        (is (= (:id appender) (:id appender')))
```
[the remark this answers is not in the document]
Comment (r0man) [reply test/clj/cider/nrepl/middleware/log_test.clj:361]: Fixed

File: test/clj/cider/nrepl/middleware/log_test.clj [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+                       :appender (:id appender)
+                       :consumer (:id consumer)
+                       :filters {:level :DEBUG}})]
+        (is (= #{"done"} (:status response)))
+        (is (= {:id (:id consumer)
+                :filters {:level "DEBUG"}}
+               (:cider/log-update-consumer response)))
+        (framework/log framework {:message "B1" :level :DEBUG})))))
+
+(defn log-something [framework & [^long n ^long sleep]]
```
[the remark this answers is not in the document]
Comment (r0man) [reply test/clj/cider/nrepl/middleware/log_test.clj:431]: I removed the hints, they don't seem to trigger a warning.

File: test/clj/cider/nrepl/middleware/log_test.clj [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+(deftest test-add-appender
+  (with-each-framework [framework (frameworks)]
+    (let [options {:filters {} :size 10 :threshold 10}
+          appender' (add-appender framework appender options)]
+      (is (= [] (:consumers appender')))
+      (is (= (:filters options) (:filters appender')))
+      (is (= (:id appender) (:id appender')))
+      (is (= (:root-logger framework) (:logger appender')))
+      (is (= (:size options) (:size appender')))
+      (is (= (:threshold options) (:threshold appender'))))
```
[the remark this answers is not in the document]
Comment (r0man) [reply test/clj/cider/nrepl/middleware/log_test.clj:100]: I added an additional check.

File: .clj-kondo/config.edn [hunk 7]
```diff
@@ -1,5 +1,6 @@
 {:hooks {:analyze-call {cider.nrepl.middleware.out/with-out-binding
                         hooks.core/with-out-binding}}
+ :lint-as {cider.nrepl.middleware.log-test/with-each-framework clojure.core/let}
```
[the remark this answers is not in the document]
Comment (r0man) [reply .clj-kondo/config.edn:3]: Do you mean to export this symbols, so it is picked up by clojure lsp in other projects?

I think we don't need this here. It's a macro defined and used only in the cider-nrepl tests. It it not used by other libraries. But maybe you mean something else?

**Comment by r0man** [comment]
> I'm fine with not inlining the dep - I don't expect any runtime conflicts with it.
> 
> Btw, how are the tests passing on Clojure 1.8? I thought logjam required 1.9.

I was also wondering when I initially opened the PR. Now that we have logjam in a separate repo, and we don't ship the specs nor reference them, it's not a hard requirement for depending on it. It's only needed for tests.

**Comment by r0man** [comment]
@vemv Thanks for the review. I hope I addressed your comments.

**Comment by bbatsov** [comment]
> I was also wondering when I initially opened the PR. Now that we have logjam in a separate repo, and we don't ship the specs nor reference them, it's not a hard requirement for depending on it. It's only needed for tests.

Ah, even better! This means that the new functionality can land in the next CIDER release!

**Pull request merged**